llpc/util/llpcElfWriter.cpp

```diff
@@ -735,6 +737,282 @@ template <class Elf> void ElfWriter<Elf>::updateMetaNote(Context *pContext, cons
+  ElfSymbol relocSym = {};
+  reader.getSymbol(inputReloc.r_symbol, &relocSym);
+
+  // Create a section named origSectionName.cached.
+  // If the section does not exist, the memory will be freed in ElfWrith destructor.
+  char *secName = new char[strlen(relocSym.secName) + strlen(CachedRodataSectionSuffix) + 1];
+  strncpy(secName, relocSym.secName, strlen(relocSym.secName) + 1);
+  strncat(secName, CachedRodataSectionSuffix, strlen(CachedRodataSectionSuffix) + 1);
+
+  for (auto section : m_sections) {
+    if (strcmp(section.name, secName) == 0) {
+      delete[] secName;
+      return;
+    }
+  }
+
+  // If the section does not exist, create a new section and update the related symbols.
+  SectionBuffer newSection = {};
+  ElfSectionBuffer<Elf64::SectionHeader> *inputRodataSection = nullptr;
+  std::vector<ElfSymbol> inputRodataSymbols;
+  auto inputRodataSecIndex = reader.GetSectionIndex(relocSym.secName);
+  reader.getSectionDataBySectionIndex(inputRodataSecIndex, &inputRodataSection);
+  reader.GetSymbolsBySectionIndex(inputRodataSecIndex, inputRodataSymbols);
+
+  createNewSection(secName, inputRodataSection, &newSection);
+
+  // Update the data from the input section.
+  // The memory will be freed in ElfWrith destructor.
+  auto secData = new uint8_t[inputRodataSection->secHead.sh_size];
+  memcpy(secData, inputRodataSection->data, inputRodataSection->secHead.sh_size);
+  newSection.data = secData;
+  newSection.secHead.sh_size = inputRodataSection->secHead.sh_size;
+
+  m_sections.push_back(newSection);
+
+  // Update symbols.
+  for (auto rodataSymbol : inputRodataSymbols) {
+    ElfSymbol symbol = {};
+    // Create a new symbol named origSymbolName_cached.
+    // The memory will be freed in ElfWrith destructor.
+    char *newSymName = new char[strlen(rodataSymbol.pSymName) + strlen(CachedRodataSymbolSuffix) + 1];
+    strncpy(newSymName, rodataSymbol.pSymName, strlen(rodataSymbol.pSymName) + 1);
+    strncat(newSymName, CachedRodataSymbolSuffix, strlen(CachedRodataSymbolSuffix) + 1);
+    symbol.secIdx = m_sections.size() - 1;
+    symbol.secName = secName;
+    symbol.size = rodataSymbol.size;
+    symbol.pSymName = newSymName;
+    symbol.value = rodataSymbol.value;
+    symbol.info.all = rodataSymbol.info.all;
+    symbol.nameOffset = InvalidValue;
+    m_symbols.push_back(symbol);
+  }
+}
+
+// =====================================================================================================================
+// Get the reloc symbol index based on the input reloc symbol
+//
+// @param inputSymbolName : The input reloc symbol name
+template <class Elf> uint32_t ElfWriter<Elf>::getRelocSymbolIndex(const char *inputSymbolName) {
+  char *newSymName = new char[strlen(inputSymbolName) + strlen(CachedRodataSymbolSuffix) + 1];
+  strncpy(newSymName, inputSymbolName, strlen(inputSymbolName) + 1);
+  strncat(newSymName, CachedRodataSymbolSuffix, strlen(CachedRodataSymbolSuffix) + 1);
```

Comment:
`strlen` is used twice on each string. Can you hoist the lengths to local variables?

Reply:
Actually, I believe that the last parameters (`n`) don't do anything because we are not passing it the right values. The remaining size in the buffer should be passed. So in the first call, this should be the size of `newSymName`, and later the same size minus `strlen(newSymName)`.

Reply:
Fixed.